File: book/swidgets/c#/SWidgets/SDateField.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Windows.Controls;
using Sodium.Frp;
using Sodium.Functional;

namespace SWidgets
{
    public class SDateField : WrapPanel
    {
        private static readonly Dictionary<string, int> Months = new Dictionary<string, int>
        {
            { "Jan", 1 },
            { "Feb", 2 },
            { "Mar", 3 },
            { "Apr", 4 },
            { "May", 5 },
            { "Jun", 6 },
            { "Jul", 7 },
            { "Aug", 8 },
            { "Sep", 9 },
            { "Oct", 10 },
            { "Nov", 11 },
            { "Dec", 12 }
        };

        public SDateField()
        {
            DateTime now = DateTime.Now;
            SComboBox<int> year = new SComboBox<int>(Maybe.Some(now.Year), Enumerable.Range(now.Year - 10, 21));
            SComboBox<string> month = new SComboBox<string>(Maybe.Some(Months.Single(p => p.Value == now.Month).Key), Months.Keys);
            SComboBox<int> day = new SComboBox<int>(Maybe.Some(now.Day), Enumerable.Range(1, 31));

            this.Children.Add(year);
            this.Children.Add(month);
            this.Children.Add(day);

            Cell<Maybe<int>> monthIndex = month.SelectedItem.Map(m => m.Match(s => Maybe.Some(Months[s]), () => Maybe.None));
            this.SelectedDate = year.SelectedItem.Lift(monthIndex, day.SelectedItem,
                (my, mm, md) =>
                    from y in my
                    from m in mm
                    from d in md
                    select new DateTime(y, m, d)).Map(m => m.Match(v => v, () => DateTime.Now));
        }

        public Cell<DateTime> SelectedDate;
    }
}

File: book/swidgets/f#/SWidgets/SDateField.fs
namespace SWidgets

open System
open System.Windows.Controls
open Sodium.Frp
open SWidgets.MaybeBuilder

type SDateField() as this =
    inherit WrapPanel()

    static let months = [("Jan",1);("Feb",2);("Mar",3);("Apr",4);("May",5);("Jun",6);("Jul",7);("Aug",8);("Sep",9);("Oct",10);("Nov",11);("Dec",12)]
    static let monthNumberByName = months |> Map.ofList
    static let monthNameByNumber = months |> List.map (fun (x,y) -> (y,x)) |> Map.ofList

    let init () =
        let now = DateTime.Now
        let year = new SComboBox<_>(Some now.Year, Seq.init 21 ((+) (now.Year - 10)))
        let month = new SComboBox<_>(Some (monthNameByNumber |> Map.find now.Month), months |> List.map fst)
        let day = new SComboBox<_>(Some now.Day, Seq.init 31 ((+) 1))
        this.Children.Add(year) |> ignore
        this.Children.Add(month) |> ignore
        this.Children.Add(day) |> ignore
        let monthIndex = month.SelectedItem |> mapC (fun o -> match o with | None -> None | Some s -> Some (monthNumberByName |> Map.find s))
        let getSelectedDate y m d =
            maybe {
                let! y = y
                let! m = m
                let! d = d
                return DateTime(y, m, d)
            }
        let dateOrNow d = match d with | None -> DateTime.Now | Some d -> d
        let selectedDate = (year.SelectedItem, monthIndex, day.SelectedItem) |> lift3C getSelectedDate |> mapC dateOrNow
        selectedDate

    let selectedDate = init ()

    member val SelectedDate = selectedDate
